### PR TITLE
Enable metric collection from arbitrary sources

### PR DESCRIPTION
**Pull request by yaacov**
**Description**

a. Current metrics collection code is Hawkular centric. This PR enable the use of arbitrary metric sources.

b. Version 4.0.0 of Hawkular client changes the error types emited on errors. This PR handle all errors, from pre and post 4.0.0 clients.

### CODE REVIEW

**Comment by yaacov** [comment]
@simon3z @cben @moolitayer @zeari please review

**Comment by simon3z** [comment]
@yaacov when this works and you want to finalize it please squash the PRs, thanks!

**Comment by yaacov** [comment]
> @yaacov when this works and you want to finalize it please squash the PRs, thanks!

Squashed 

catnp

**Comment by moolitayer** [comment]
@yaacov where are these new files used?
```
hawkular_capture_context_container_metrics.yml
hawkular_capture_context_node_metrics.yml
hawkular_capture_context_pod_metrics.yml
```

**Comment by moolitayer** [comment]
@yaacov why did you rename yaml files `s/_metric/_timespan`?

Generally look good @yaacov thanks for splitting the #20 let's get another person to review this
@cben

File: app/models/manageiq/providers/kubernetes/container_manager/metrics_capture/capture_context_mixin.rb [hunk 1]
```diff
@@ -56,77 +54,12 @@ def validate_target
...
-          :starts         => @starts,
-          :ends           => @ends,
-          :bucketDuration => "#{@interval}s"))
-    rescue SystemCallError, SocketError, OpenSSL::SSL::SSLError => e
-      raise CollectionFailure, e.message
-    end
-
     def process_cpu_counters_rate(counters_rate)
-      @metrics |= ['cpu_usage_rate_average'] if counters_rate.length > 0
+      @metrics |= ['cpu_usage_rate_average'] unless counters_rate.empty?
```
Comment (cben) [inline app/models/manageiq/providers/kubernetes/container_manager/metrics_capture/capture_context_mixin.rb]: orthogonal to this PR: `sort_and_normalize` discards last entry => maybe these conditions should be `>= 1` ?

**Comment by simon3z** [comment]
@cben @moolitayer this PR is lacking approvals. Still reviewing?

**Comment by moolitayer** [comment]
> @yaacov why did you rename yaml files s/_metric/_timespan?

I think I get it now, @yaacov please check if you can safely remove these files:
```
capture_context_container_timespan.yml
capture_context_node_timespan.yml
capture_context_pod_timespan.yml
```

**Comment by yaacov** [comment]
> @yaacov please check if you can safely remove these files

Nice catch 👍 removed unneeded files

File: app/models/manageiq/providers/kubernetes/container_manager/metrics_capture/capture_context_mixin.rb [hunk 2]
```diff
@@ -1,12 +1,10 @@
...
-  class CaptureContext
-    include ManageIQ::Providers::Kubernetes::ContainerManager::MetricsCapture::HawkularClientMixin
-
+  module CaptureContextMixin
     def initialize(target, start_time, end_time, interval)
       @target = target
       @starts = start_time.to_i.in_milliseconds
       @ends = end_time.to_i.in_milliseconds if end_time
-      @interval = interval
+      @interval = interval.to_i
```
Comment (moolitayer) [inline app/models/manageiq/providers/kubernetes/container_manager/metrics_capture/capture_context_mixin.rb]: @yaacov ~~will the existing hawkular implementation work fine with this change ?~~
OK usage in `:bucketDuration` and `starts` seems to be unchanged as far as I can tell

**Comment by simon3z** [comment]
@cben @moolitayer so we're blocked on the Travis issue?
cc @Fryguy

**Comment by moolitayer** [comment]
> @cben @moolitayer so we're blocked on the Travis issue?
> cc @Fryguy

Yes

**Comment by cben** [comment]
We could checkout pulls/NN/merge, test  locally & merge...

On Jul 18, 2017 12:41, "Mooli Tayer" <notifications@github.com> wrote:

@cben <https://github.com/cben> @moolitayer <https://github.com/moolitayer>
so we're blocked on the Travis issue?
cc @Fryguy <https://github.com/fryguy>

Yes

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/62#issuecomment-316012530>,
or mute the thread

**Comment by simon3z** [comment]
@cben can you review/approve? thanks.

File: app/models/manageiq/providers/kubernetes/container_manager/metrics_capture/hawkular_capture_context.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      process_net_counters_rate(compute_summation(net_counters))
+    end
+
+    def fetch_counters_data(resource)
+      sort_and_normalize(
+        hawkular_client.counters.get_data(
+          resource,
+          :starts         => @starts - @interval.in_milliseconds,
+          :ends           => @ends,
+          :bucketDuration => "#{@interval}s"
```
Comment (cben) [inline app/models/manageiq/providers/kubernetes/container_manager/metrics_capture/hawkular_capture_context.rb]: so interval is in seconds, starts & ends in ms?
it's impossible to know looking at this code.
can we rename vars to `@starts_ms` etc?

**Comment by cben** [comment]
LGTM :+1:
My comment about units is out of scope, as this is moved code.

**Pull request merged**